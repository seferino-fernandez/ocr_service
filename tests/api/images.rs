use axum::{
    body::Body,
    http::{Request, StatusCode, header::CONTENT_TYPE},
};
use http_body_util::BodyExt as _;
use tokio::fs::read;

use crate::helpers::*;

#[tokio::test]
async fn test_images_endpoint_tessdoc_introduction() {
    let app = TestApp::new();

    // Read test image file
    let image_data = read("tests/images/tessdoc-introduction.png").await.unwrap();

    // Create multipart form data
    let body = create_multipart_body("image", "tessdoc-introduction.png", &image_data);

    let req = Request::post("/api/v1/images")
        .header(
            CONTENT_TYPE,
            format!("multipart/form-data; boundary={}", BOUNDARY),
        )
        .body(body)
        .unwrap();

    let response = app.request(req).await;
    assert_eq!(response.status(), StatusCode::OK);

    let body = response.into_body().collect().await.unwrap().to_bytes();
    let body: serde_json::Value = serde_json::from_slice(&body).unwrap();
    insta::with_settings!({filters => vec![
        (r"\\n\*", r"\ne"),
    ]}, {
        insta::assert_yaml_snapshot!(body);
    });
}

// Helper constant for multipart boundary
const BOUNDARY: &str = "test_boundary";

// Helper function to create multipart form data
fn create_multipart_body(field_name: &str, filename: &str, data: &[u8]) -> Body {
    let mut body = Vec::new();

    // Add multipart boundary and headers
    body.extend_from_slice(format!("--{}\r\n", BOUNDARY).as_bytes());
    body.extend_from_slice(
        format!(
            "Content-Disposition: form-data; name=\"{}\"; filename=\"{}\"\r\n",
            field_name, filename
        )
        .as_bytes(),
    );
    body.extend_from_slice("Content-Type: image/png\r\n\r\n".as_bytes());

    // Add file data
    body.extend_from_slice(data);
    body.extend_from_slice(format!("\r\n--{}--\r\n", BOUNDARY).as_bytes());

    Body::from(body)
}
