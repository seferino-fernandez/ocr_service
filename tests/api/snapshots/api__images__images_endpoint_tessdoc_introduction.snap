---
source: tests/api/images.rs
expression: body
---
text: "Introduction\nTesseract is an open source text recognition (OCR) Engine, available under the Apache 2.0 license.\ne Major version 5 is the current stable version and started with release 5.0.0 on November 30,\n2021.\ne Newer minor versions and bugfix versions are available from GitHub.\ne Latest source code is available from main branch on GitHub. Open issues can be found in issue\ntracker, and planning documentation.\nTesseract can be used directly via command line, or (for programmers) by using an API to extract\nprinted text from images. It supports a wide variety of languages. Tesseract doesn't have a built-in\nGUI, but there are several available from the 3rdParty page. External tools, wrappers and training\nprojects for Tesseract are listed under AddOns.\nTesseract can be used in your own project, under the terms of the Apache License 2.0. It has a fully\nfeatured API, and can be compiled for a variety of targets including Android and the iPhone. See the\n3rdParty and AddOns pages for samples of what has been done with it.\nIf you have a question, first read the documentation, particularly the FAQ to see if your problem is\naddressed there. If not, search the Issues List, Tesseract user forum, and if you still can't find what\nyou need, please ask your question in Tesseract user forum Google group.\nTesseract is free software, so if you want to pitch in and help, please do! If you find a bug and fix it\nyourself, the best thing to do is to attach the patch to your bug report in the Issues List.\n"
